W-11862525-first-draft-for-beta-site-build-CP (#114)
mulesoft/docs-apikit: modules/ROOT/pages/apikit-odatav4-expand-feature.adoc

= Enhancing API Consumption and Minimizing Network Traffic
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> You can reduce the amount of incoming network traffic and make APIs easier to consume with the `$expand` system query option. Because API consumers request only inline representations of the related entities of query results, the returned information avoids multiple network roundtrips. This benefits metered and latency-sensitive consumers.

Reviewer: ```suggestion
To reduce the amount of incoming network traffic and make APIs easier to consume, use the `$expand` system query option. Because API consumers request only inline representations of the related entities of query results, the returned information avoids multiple network roundtrips. Using this query option benefits metered and latency-sensitive consumers.
```

Author: OK! Thanks!

* The `Employee` entity has a related `Manager` entity.
* The `Employees` entity set is a collection of `Employee` entities.
* The `Managers` entity set is a collection of `Manager` entities.
* The API location is `+https://example.com/api/+`.

Given this setup, the following requests retrieve employees and their managers:

* Use `+GET https://example.com/api/Employees+` to retrieve a list of employees.
* Gather the `Manager` IDs through the `ReferentialConstraint` of their relationship.
* Use `+GET https://example.com/api/Managers?$filter=ManagerID in (id1, id2, ..., idN)+` to retrieve a list of related managers.
* Write a small program to merge the lists based on join criteria.

> If using the `$expand` feature, request is `+GET https://example.com/api/Employees?$expand=Manager+` to retrieve a list of employees with their `Manager` navigation property expanded.

Reviewer 2: ```suggestion
If using the `$expand` feature, the request` to retrieve a list of employees with their `Manager` navigation property expanded is `+GET https://example.com/api/Employees?$expand=Manager+.
```

* The `Employee` entity has a related `Manager` entity.
* The `Employees` entity set is a collection of `Employee` entities.
* The `Managers` entity set is a collection of `Manager` entities.
* The API location is `+https://example.com/api/+`.

Given this setup, the following requests retrieve employees and their managers:

* Use `+GET https://example.com/api/Employees+` to retrieve a list of employees.
* Gather the `Manager` IDs through the `ReferentialConstraint` of their relationship.
* Use `+GET https://example.com/api/Managers?$filter=ManagerID in (id1, id2, ..., idN)+` to retrieve a list of related managers.
* Write a small program to merge the lists based on join criteria.

> If using the `$expand` feature, the request to retrieve a list of employees with their `Manager` navigation property expanded is `+GET https://example.com/api/Employees?$expand=Manager+`.

Reviewer: ```suggestion
If you use the `$expand` feature, the request to retrieve a list of employees with their `Manager` navigation property expanded is `+GET https://example.com/api/Employees?$expand=Manager+`.
```

Author: OK! Thanks!

== The `generic expand` Feature

> You can reuse existing flow implementations as a starting point for expanding requests by using the `generic expand` feature. You can focus on the relevant data retrieval operations while the module performs the necessary steps to ensure that the `$expand` fields are complete.

Reviewer: ```suggestion
Reuse existing flow implementations as a starting point for expanding requests by using the `generic expand` feature, which enables you to focus on the relevant data retrieval operations while the module performs the necessary steps to ensure that the `$expand` fields are complete.
```

Author: OK! Thanks!

The `generic expand` feature dispatches additional queries to the application when the original flow completes. Because this is done based on the application `EDM`, you must make extra annotations to add enough context to the module:

* The `EntitySet` where the expansion takes place has a `NavigationPropertyBinding` for every expandable `NavigationProperty` of its `Entity`. 
+
This is used to select which `<entity-collection-source/>` is executed when expanding that property.
* Each expandable `NavigationProperty` has a `ReferentialConstraint` attached. 
+

> Internal requests use it when the value of `Property` is matched against the value of `ReferencedProperty`.

Reviewer: ```suggestion
Internal requests use `ReferentialConstraint` when matching the value of `Property` against the value of `ReferencedProperty`.
```
It's a bit unclear what "it" refers to in this case. Please confirm if the suggestion is correct.

Author: Yes! You're right! Thanks!
